source/common/ssl/context_impl.cc

```diff
@@ -131,6 +131,15 @@ ContextImpl::ContextImpl(ContextManagerImpl& parent, Stats::Scope& scope,
     verify_mode = SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT;
   }
 
+  // NOTE: We're using SSL_CTX_set_cert_verify_callback() instead of X509_verify_cert()
+  // directly. However, our new callback is still calling X509_verify_cert() under
+  // the hood. Therefore, to ignore cert expiration, we need to set the callback
+  // for X509_verify_cert to ignore that error.
+  if (config.disableCertificateExpirationVerification()) {
+    X509_STORE* store = SSL_CTX_get_cert_store(ctx_.get());
+    X509_STORE_set_verify_cb(store, ContextImpl::verifyIgnoreCertExpirationCallback);
+  }
```

Comment:
It's ugly, but I think necessary since `ContextImpl::verifyCallback()` is piggybacking on `X509_verify_cert()` for part of the validation.

`X509_verify_cert` calls `ctx->verify` (or the default `internal_verify`) which loops through the chain.  If an iteration fails due to an expiration in `internal_verify`, it will abort the loop and exit early which then cascades as `X509_verify_cert` also exits early on error.  If you check for expiration as an error condition from `X509_verify_cert` rather than using a callback to ignore it, you're skipping over other validation checks.